app/uk/gov/hmrc/plasticpackagingtax/returns/controllers/actions/AuthAgentAction.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.plasticpackagingtax.returns.controllers.actions
+
+import com.google.inject.{ImplementedBy, Inject}
+import com.kenshoo.play.metrics.Metrics
+import play.api.Logger
+import play.api.mvc._
+import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals._
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.plasticpackagingtax.returns.config.AppConfig
+import uk.gov.hmrc.plasticpackagingtax.returns.controllers.home.{routes => homeRoutes}
+import uk.gov.hmrc.plasticpackagingtax.returns.models.SignedInUser
+import uk.gov.hmrc.plasticpackagingtax.returns.models.request.{AuthenticatedRequest, IdentityData}
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthAgentActionImpl @Inject() (
+  override val authConnector: AuthConnector,
+  appConfig: AppConfig,
+  metrics: Metrics,
+  mcc: MessagesControllerComponents
+) extends AuthAgentAction with AuthorisedFunctions {
+
+  implicit override val executionContext: ExecutionContext = mcc.executionContext
+  override val parser: BodyParser[AnyContent]              = mcc.parsers.defaultBodyParser
+  private val logger                                       = Logger(this.getClass)
+  private val authTimer                                    = metrics.defaultRegistry.timer("ppt.returns.upstream.auth.timer")
+
+  private val authData =
+    credentials and name and email and externalId and internalId and affinityGroup and allEnrolments and
+      agentCode and confidenceLevel and nino and saUtr and dateOfBirth and agentInformation and groupIdentifier and
+      credentialRole and mdtpInformation and itmpName and itmpDateOfBirth and itmpAddress and credentialStrength and loginTimes
+
+  override def invokeBlock[A](
+    request: Request[A],
+    block: AuthenticatedRequest[A] => Future[Result]
+  ): Future[Result] = {
+    implicit val hc: HeaderCarrier =
+      HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+    val authorisation = authTimer.time()
+
+    authorised(AffinityGroup.Agent)
+      .retrieve(authData) {
+        case credentials ~ name ~ email ~ externalId ~ internalId ~ affinityGroup ~ allEnrolments ~ agentCode ~
```

Comment:
Optional - could some of these retrievals be extracted and shared with AuthAction where they're replicated?

Reply:
Almost certainly.
There are elements of merge conflict fear and Scala fear here.
I'd like to get merged, validate this and then look for some clean ups if that's ok.

Reply:
yes, that's fine.